src/plugins/dataflow-tool/microbit/hub.md

```diff
@@ -2,61 +2,111 @@
 
 The program for the micro:bit attached to the sensors/relays
 
-Every two seconds
-  - Get a temp reading
-  - Get a humidity reading
-  - Make a packet for temp
-  - Make a packet for humidity
-  - Send temp packet
-  - Send humidity packet
+Downloadable link: https://makecode.microbit.org/_YpoCfzeTwCWP
 
+#### Outgoing strings to write,
+e.g. : `sat20.31`
+- `s`: "sensor"
+- `a`: "hub A"
+- `t`: "temperature reading"
+- `<n>`: the reading as a string of num chars
 
-On radio recieved
-  - If message type == `c` and addressed to my id
-  - Change relay states accordingly
-  - Send confirmation of relay states back as an `s`
+#### Incoming strings to read
+e.g. : `ca11`
+- `c`: "control"
+- `a`: "hub A"
+- `1`: "relay at index 1"
+- `1`: "turn on"
 
+#### Pins
+- `15`    dht sensor (should change to `16`?):
+- `8`     relay 2 (sprinkler) (green)
+- `9`     relay 1 (fan) (yellow)
+- `12`    relay 0 (heat) (red)
 
 ```js
 radio.setGroup(1)
 
-let hubIds = ['x','a','b','c','d']
+let hubIds = ['x', 'a', 'b', 'c', 'd']
 let hubI = 0
-let readingInterval = 2000
-let messsageInterval = 100
 
-function getTempString(){
-  const t = Math.constrain(dht11_dht22.readData(dataType.temperature), 0, 100);
-  return `s${hubIds[hubI]}t${t}`;
+// in order to map to relayIndex
+let relayPins = [DigitalPin.P12, DigitalPin.P9, DigitalPin.P8]
+
+function getTempString() {
+    const t = Math.constrain(dht11_dht22.readData(dataType.temperature), 0, 100);
+    return `s${hubIds[hubI]}t${t}`;
 }
 
-function getHumidString(){
-  const h = Math.constrain(dht11_dht22.readData(dataType.humidity), 0, 100);
-  return `s${hubIds[hubI]}h${h}`;
+function getHumidString() {
+    const h = Math.constrain(dht11_dht22.readData(dataType.humidity), 0, 100);
+    return `s${hubIds[hubI]}h${h}`;
 }
 
 function readSendData() {
-  dht11_dht22.queryData(DHTtype.DHT11, DigitalPin.P15, false, false, false)
-  radio.sendString(getTempString())
-  pause(messageInterval)
-  radio.sendString(getHumidString())
+    dht11_dht22.queryData(DHTtype.DHT11, DigitalPin.P15, false, false, false)
+    radio.sendString(getTempString())
+    pause(1000) //readings improve with this delay
+    radio.sendString(getHumidString())
+}
+
+function operateRelay(relayIndex: number, state: number) {
+
+    //operate the relay
+    pins.digitalWritePin(relayPins[relayIndex], state)
```

Comment:
The `relayIndex` should be validated here. If some bogus radio signal comes in with an invalid relayIndex that will cause the program to read out of the array bounds. That might crash the program by itself, or it might cause digitalWritePin to get a pin value that is weird which could wreak havoc on the microbit.

Reply:
I ended using a less-than-elegant validation - I don't believe the implementation of JavaScript used to author the micro:bit programs supports regex or `Array.includes`.  

```js
    const validRelay = relayIndex === 0 || relayIndex === 1 || relayIndex === 2;
    const validRelaySignal = state === 0 || state === 1;
    if (validRelay && validRelaySignal){
        pins.digitalWritePin(relayPins[relayIndex], state)
    }
```

Reply:
Seems fine to me. 
You could move the `plot` line that follows into the if, and possible show some symbol for "error" when they are not valid. But none of that is critical.